clippy_lints/src/operators/float_cmp.rs

```diff
@@ -1,47 +1,87 @@
-            None => true,
-        };
+    let peel_expr = |e: &'tcx Expr<'tcx>| match e.kind {
+        ExprKind::Cast(e, _) | ExprKind::AddrOf(BorrowKind::Ref, _, e) | ExprKind::Unary(UnOp::Neg, e) => Some(e),
+        _ => None,
+    };
 
+    if matches!(op, BinOpKind::Eq | BinOpKind::Ne)
+        && let left_reduced = peel_hir_expr_while(left, peel_expr)
+        && let right_reduced = peel_hir_expr_while(right, peel_expr)
+        && is_float(cx, left_reduced)
+        // Don't lint literal comparisons
+        && !(matches!(left_reduced.kind, ExprKind::Lit(_)) && matches!(right_reduced.kind, ExprKind::Lit(_)))
         // Allow comparing the results of signum()
-        if is_signum(cx, left) && is_signum(cx, right) {
+        && !(is_signum(cx, left_reduced) && is_signum(cx, right_reduced))
+        && match (path_res(cx, left_reduced), path_res(cx, right_reduced)) {
+            (Res::Err, _) | (_, Res::Err) => true,
+            (left, right) => left != right,
+        }
+    {
+        let left_c = constant(cx, cx.typeck_results(), left_reduced);
+        let is_left_const = left_c.is_some();
+        if left_c.is_some_and(|c| is_allowed(&c)) {
+            return;
+        }
+        let right_c = constant(cx, cx.typeck_results(), right_reduced);
+        let is_right_const = right_c.is_some();
+        if right_c.is_some_and(|c| is_allowed(&c)) {
+            return;
+        }
+
+        if config.ignore_constant_comparisons
+            && (is_left_const || is_const_evaluatable(cx, left_reduced))
+            && (is_right_const || is_const_evaluatable(cx, right_reduced))
+        {
+            return;
+        }
+
+        if config.ignore_named_constants
+            && (is_expr_named_const(cx, left_reduced) || is_expr_named_const(cx, right_reduced))
+        {
+            return;
+        }
+
+        if config.ignore_change_detection
+            && ((is_pure_expr(cx, left_reduced) && contains_expr(cx, right, left))
+                || (is_pure_expr(cx, right_reduced) && contains_expr(cx, left, right)))
+        {
             return;
         }
 
         if let Some(name) = get_item_name(cx, expr) {
             let name = name.as_str();
-            if name == "eq" || name == "ne" || name == "is_nan" || name.starts_with("eq_") || name.ends_with("_eq") {
+            if name == "eq" || name == "ne" || name.starts_with("eq_") || name.ends_with("_eq") {
                 return;
             }
         }
-        let is_comparing_arrays = is_array(cx, left) || is_array(cx, right);
-        let (lint, msg) = get_lint_and_message(left_is_local && right_is_local, is_comparing_arrays);
-        span_lint_and_then(cx, lint, expr.span, msg, |diag| {
+        let is_comparing_arrays = is_array(cx, left_reduced) || is_array(cx, right_reduced);
+        let msg = if is_comparing_arrays {
+            "strict comparison of `f32` or `f64` arrays"
+        } else {
+            "strict comparison of `f32` or `f64`"
+        };
```

Comment:
Optional NIT:
I think it would be cool to not have "`f32` or `f64`" but the direct type. At this point the lint will be emitted, so it makes sense to "spent" a function call to get the type and print it in the message IMO.

Reply:
I don't see the type even being useful here, but didn't really want to deal with rewording it. Personally I think just "a floating point type" is a good enough message.

I really want to go through and fix up all of out error messages since a lot of them are questionable, but that's a pretty big task.